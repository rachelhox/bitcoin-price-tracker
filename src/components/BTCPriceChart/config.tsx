export const customPeriodKeys: number[] = [30, 7, 1];
