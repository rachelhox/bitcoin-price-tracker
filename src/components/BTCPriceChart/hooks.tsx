/* eslint-disable no-undef */
/* eslint-disable array-callback-return */
/* eslint-disable no-console */
import { useState, useEffect } from 'react';
import axios from 'axios';
import useSWR from 'swr';
import { useGeneralContext } from '@contexts';

export const useLineChartkHook = () => {
  const { currency } = useGeneralContext();
  const [lineChartData, setLineChartData] = useState<Array<PriceProps>>([]);
  const [period, setPeriod] = useState(30);
  const fetcher = (url: string) => axios.get(url).then((res) => res.data);
  // eslint-disable-next-line no-unused-vars
  const { data: marketData, isValidating } = useSWR(
    `${process.env.NEXT_PUBLIC_COINGECKO_API_URL}/coins/bitcoin/market_chart?vs_currency=${currency.key}&days=${period}`,
    fetcher,
    { refreshInterval: 60000 }
  );

  useEffect(() => {
    setLineChartData([]);
    if (marketData) {
      marketData.prices?.map((line: any) => {
        const dateObj = new Date(line[0]);
        const day = dateObj.getDate();
        const month = dateObj.getMonth() + 1;
        const year = dateObj.getFullYear();
        const hour = dateObj.getHours();
        const minute = dateObj.getMinutes();
        setLineChartData((prev) => [
          ...prev,
          {
            x: new Date(year, month, day, hour, minute),
            y: line[1],
          },
        ]);
      });
    }
  }, [marketData, isValidating, currency, period]);

  return { currency, period, setPeriod, lineChartData };
};
