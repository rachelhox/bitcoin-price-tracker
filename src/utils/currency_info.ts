interface IStringIndex extends Record<string, any> {}

/**
 * Takes the key name and returns the location. Will return null if undefined
 */

const currencies: IStringIndex = {
  usd: {
    name: 'US Dollar',
    key: 'usd',
    unit: '$',
    type: 'fiat',
  },
  aed: {
    name: 'United Arab Emirates Dirham',
    key: 'aed',
    unit: 'DH',
    type: 'fiat',
  },
  eur: {
    name: 'Euro',
    key: 'eur',
    unit: '€',
    type: 'fiat',
  },
  aud: {
    name: 'Australian Dollar',
    key: 'aud',
    unit: 'A$',
    type: 'fiat',
  },
  cad: {
    name: 'Canadian Dollar',
    key: 'cad',
    unit: 'CA$',
    type: 'fiat',
  },
  chf: {
    name: 'Swiss Franc',
    key: 'chf',
    unit: 'Fr.',
    type: 'fiat',
  },
  gbp: {
    name: 'British Pound Sterling',
    key: 'gbp',
    unit: '£',
    type: 'fiat',
  },
  hkd: {
    name: 'Hong Kong Dollar',
    key: 'hkd',
    unit: 'HK$',
    type: 'fiat',
  },
  sgd: {
    name: 'Singapore Dollar',
    key: 'sgd',
    unit: 'S$',
    type: 'fiat',
  },
  cny: {
    name: 'Chinese Yuan',
    key: 'cny',
    unit: '¥',
    type: 'fiat',
  },
  jpy: {
    name: 'Japanese Yen',
    key: 'jpy',
    unit: '¥',
    type: 'fiat',
  },
  inr: {
    name: 'Indian Rupee',
    key: 'inr',
    unit: '₹',
    type: 'fiat',
  },
  krw: {
    name: 'South Korean Won',
    key: 'krw',
    unit: '₩',
    type: 'fiat',
  },
};

export const getCurrencyInfo = (key: string | number) => {
  return currencies[key] ?? {};
};
