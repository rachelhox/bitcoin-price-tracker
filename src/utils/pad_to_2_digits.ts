/* eslint-disable */

export const padTo2Digits = (num: any) => {
  return String(num).padStart(2, '0');
};
