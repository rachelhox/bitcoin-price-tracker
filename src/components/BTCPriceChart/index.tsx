/* eslint-disable no-unused-vars */
import {
  Box,
  Tooltip,
  Typography,
  IconButton,
  useTheme,
  useMediaQuery,
} from '@mui/material';
import InfoOutlinedIcon from '@mui/icons-material/InfoOutlined';
import { TimePeriodButton } from '@components';
import useStyles from './useStyles';
import { LineChart } from './components';
import { useLineChartHook } from './hooks';
import { customPeriodKeys } from './config';

const BTCPriceChart = () => {
  const styles = useStyles();
  const theme = useTheme();
  const { currency, period, setPeriod, lineChartData } = useLineChartHook();
  const onlyLargeScreen = useMediaQuery(theme.breakpoints.up('laptop'));

  return (
    <Box css={styles.root}>
      <Box css={styles.maxWidth}>
        <Box css={styles.container}>
          <Box css={styles.title}>
            <Typography variant="h1" css={styles.h1}>
              {`BTC/${currency.key.toUpperCase()}`}
            </Typography>
            {onlyLargeScreen ? (
              <Tooltip title="Drag and zoom in on the chart and hover to see each data point">
                <IconButton>
                  <InfoOutlinedIcon color="primary" />
                </IconButton>
              </Tooltip>
            ) : (
              <Typography variant="h3" css={styles.h3}>
                *Drag and zoom in on the chart and tap to see each data point
              </Typography>
            )}
          </Box>
          <TimePeriodButton
            period={period}
            setPeriod={setPeriod}
            customPeriod={customPeriodKeys}
          />
        </Box>
        <LineChart lineChartData={lineChartData} />
      </Box>
    </Box>
  );
};

export default BTCPriceChart;
