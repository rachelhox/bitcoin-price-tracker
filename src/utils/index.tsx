import { getCurrencyInfo } from './currency_info';
import { convertToMoney } from './convert_to_money';
import { padTo2Digits } from './pad_to_2_digits';

export { getCurrencyInfo, convertToMoney, padTo2Digits };
