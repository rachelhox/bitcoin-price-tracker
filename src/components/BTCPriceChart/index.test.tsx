import renderer from 'react-test-renderer';
import { MockTheme } from '@tests/utils';
import BTCPriceChart from '.';

// ==================================
// mocks
// ==================================

jest.mock('@components', () => ({
  TimePeriodButton: (props: any) => <div id="TimePeriodButton" {...props} />,
}));

jest.mock('./components', () => ({
  LineChart: (props: any) => <div id="LineChart" {...props} />,
}));

// ==================================
// unit tests
// ==================================
describe('component: BTCPriceChart', () => {
  it('matches snapshot', () => {
    const component = renderer.create(
      <MockTheme>
        <BTCPriceChart />
      </MockTheme>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
