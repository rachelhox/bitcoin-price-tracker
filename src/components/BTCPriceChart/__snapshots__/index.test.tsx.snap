// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: BTCPriceChart matches snapshot 1`] = `
<div
  className="MuiBox-root css-t16hmq-BTCPriceChart"
>
  <div
    className="MuiBox-root css-4yzvp0-BTCPriceChart"
  >
    <div
      className="MuiBox-root css-7dv8nf-BTCPriceChart"
    >
      <div
        className="MuiBox-root css-5d66mx-BTCPriceChart"
      >
        <h1
          className="MuiTypography-root MuiTypography-h1 css-ahnsnm-MuiTypography-root-BTCPriceChart"
        >
          BTC/USD
        </h1>
        <h3
          className="MuiTypography-root MuiTypography-h3 css-1hszivs-MuiTypography-root-BTCPriceChart"
        >
          *Tap to see each data point on the chart
        </h3>
      </div>
      <div
        customPeriod={
          [
            30,
            7,
            1,
          ]
        }
        id="TimePeriodButton"
        period={30}
        setPeriod={[Function]}
      />
    </div>
    <div
      id="LineChart"
      lineChartData={[]}
    />
  </div>
</div>
`;
